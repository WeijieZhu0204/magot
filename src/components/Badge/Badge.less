@import '../../variable.less';

.@{prefix}-badge {
  .box-sizing();
  .box();

  position: relative;
  display: inline-block;
}

.@{prefix}-badge-count {
  padding: @badge-padding;
  min-width: @badge-min-width;
  height: @badge-height;
  line-height: @badge-height;
  color: @badge-color;
  font-size: @badge-font-size;
  font-weight: normal;
  font-style: normal;
  text-align: center;
  white-space: nowrap;
  overflow: hidden;
  border-radius: @badge-border-radius;
  vertical-align: baseline;
}

.@{prefix}-badge-dot {
  width: @badge-dot-size;
  height: @badge-dot-size;
  border-radius: 50%;
}

.@{prefix}-badge-count,
.@{prefix}-badge-dot {
  position: absolute;
  top: 0;
  right: 0;
  transform: translate(50%, -50%);
  transform-origin: right 0;
  background-color: @badge-background-color;
  box-shadow: @badge-box-shadow;
  z-index: @badge-zindex;
  .box-sizing();
}

.@{prefix}-badge-empty {
  vertical-align: middle;
  .@{prefix}-badge-count,
  .@{prefix}-badge-dot {
    position: relative;
    display: block;
    transform: none;
  }
}
