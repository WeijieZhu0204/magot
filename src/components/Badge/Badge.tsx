import React from 'react';

import './Badge.less';

import * as component from '../component';

export interface BadgeProps
  extends component.BaseComponent,
    component.NestedComponent {
  count?: number;
  max?: number;
  dot?: boolean;
}

const defaultProps: Partial<BadgeProps> = {
  count: 0,
  max: 99,
  dot: false,
};

function getDisplayCount(count: number, max: number) {
  if (count <= max) return '' + count;
  return max + '+';
}

function Badge(props: BadgeProps) {
  const type = 'badge';
  const prefix = component.getComponentPrefix(type);
  const cls = component.getComponentClasses(type, props, {
    [prefix + '-empty']: !props.children,
  });
  return (
    <div className={cls}>
      {props.children}
      {props.count! > 0 && (
        <sup
          className={prefix + (props.dot ? '-dot' : '-count')}
          title={'' + props.count}
          style={props.style}>
          {props.dot ? '' : getDisplayCount(props.count!, props.max!)}
        </sup>
      )}
    </div>
  );
}

Badge.defaultProps = defaultProps;

export default Badge;
