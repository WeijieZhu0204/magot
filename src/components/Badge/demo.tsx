import React from 'react';
import Badge from './Badge';
import { Link } from 'react-router-dom';

function BadgeDemo() {
  return (
    <>
      <Link to="/" className="demo-goback">
        返回
      </Link>
      <div className="demo-box">
        <Badge count={5}>
          <a href="#" className="demo-card" />
        </Badge>
        <Badge count={10}>
          <a href="#" className="demo-card" />
        </Badge>
        <Badge count={100}>
          <a href="#" className="demo-card" />
        </Badge>
        <Badge count={999} max={999}>
          <a href="#" className="demo-card" />
        </Badge>
        <Badge count={1000} max={999}>
          <a href="#" className="demo-card" />
        </Badge>
        <Badge count={1000} max={999} style={{ backgroundColor: '#52c41a' }}>
          <a href="#" className="demo-card" />
        </Badge>
        <Badge count={5} dot={true}>
          <a href="#" className="demo-card" />
        </Badge>
      </div>
      <div className="demo-box">
        <Badge count={5} />
        <Badge count={10} />
        <Badge count={100} />
        <Badge count={999} max={999} />
        <Badge count={1000} max={999} />
        <Badge count={1000} max={999} style={{ backgroundColor: '#52c41a' }} />
        <Badge count={5} dot={true} />
      </div>
    </>
  );
}

export default BadgeDemo;
