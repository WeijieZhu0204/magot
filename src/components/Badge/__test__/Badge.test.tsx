import React from 'react';
import { render } from 'enzyme';
import Badge from '../Badge';

describe('Badge', () => {
  it('should renders dom correctly', () => {
    const wrapper = render(<Badge count={10} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should renders when count is 0', () => {
    const wrapper = render(<Badge count={0} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('sup')).toHaveLength(0);
  });

  it('should renders `count` when pass in', () => {
    const wrapper = render(<Badge count={10} />);
    const node = wrapper.find('.mgt-badge-count');
    expect(wrapper).toMatchSnapshot();
    expect(node).toHaveLength(1);
    expect(node.text()).toEqual('10');
    expect(node.attr('title')).toEqual('10');
  });

  it('should renders large `count` when pass in', () => {
    const wrapper = render(<Badge count={100} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.mgt-badge-count').text()).toEqual('99+');
  });

  it('should renders `count` and `max` when pass in', () => {
    const wrapper = render(<Badge count={100} max={1000} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.mgt-badge-count').text()).toEqual('100');
  });

  it('should renders `dot` when pass in', () => {
    const wrapper = render(<Badge count={5} dot={true} />);
    const node = wrapper.find('.mgt-badge-dot');
    expect(wrapper).toMatchSnapshot();
    expect(node).toHaveLength(1);
    expect(node.text()).toEqual('');
  });

  it('should renders when wraps other dom', () => {
    const wrapper = render(
      <Badge count={5}>
        <a href="#" />
      </Badge>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('a')).toHaveLength(1);
    expect(wrapper.hasClass('.mgt-badge-empty')).toBe(false);
  });

  it('should renders custom className', () => {
    const wrapper = render(<Badge className="jest-test-class" />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.hasClass('jest-test-class')).toEqual(true);
  });

  it('should renders custom styles', () => {
    const wrapper = render(
      <Badge count={5} style={{ color: '#333', backgroundColor: '#f0f0f0' }} />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('sup').attr('style')).toBeDefined();
  });
});
