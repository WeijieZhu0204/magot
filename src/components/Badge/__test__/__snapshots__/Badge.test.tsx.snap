// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should renders \`count\` and \`max\` when pass in 1`] = `
<div
  class="mgt-badge mgt-badge-empty"
>
  <sup
    class="mgt-badge-count"
    title="100"
  >
    100
  </sup>
</div>
`;

exports[`Badge should renders \`count\` when pass in 1`] = `
<div
  class="mgt-badge mgt-badge-empty"
>
  <sup
    class="mgt-badge-count"
    title="10"
  >
    10
  </sup>
</div>
`;

exports[`Badge should renders \`dot\` when pass in 1`] = `
<div
  class="mgt-badge mgt-badge-empty"
>
  <sup
    class="mgt-badge-dot"
    title="5"
  />
</div>
`;

exports[`Badge should renders custom className 1`] = `
<div
  class="mgt-badge mgt-badge-empty jest-test-class"
/>
`;

exports[`Badge should renders custom styles 1`] = `
<div
  class="mgt-badge mgt-badge-empty"
>
  <sup
    class="mgt-badge-count"
    style="color:#333;background-color:#f0f0f0"
    title="5"
  >
    5
  </sup>
</div>
`;

exports[`Badge should renders dom correctly 1`] = `
<div
  class="mgt-badge mgt-badge-empty"
>
  <sup
    class="mgt-badge-count"
    title="10"
  >
    10
  </sup>
</div>
`;

exports[`Badge should renders large \`count\` when pass in 1`] = `
<div
  class="mgt-badge mgt-badge-empty"
>
  <sup
    class="mgt-badge-count"
    title="100"
  >
    99+
  </sup>
</div>
`;

exports[`Badge should renders when count is 0 1`] = `
<div
  class="mgt-badge mgt-badge-empty"
/>
`;

exports[`Badge should renders when wraps other dom 1`] = `
<div
  class="mgt-badge"
>
  <a
    href="#"
  />
  <sup
    class="mgt-badge-count"
    title="5"
  >
    5
  </sup>
</div>
`;
